// dojo-widgets/dist/umd/util/lang.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    function getIndex(list, item, position, reference) {
        var idx;
        if (typeof position === 'number') {
            idx = position;
            var size = Array.isArray(list) ? list.length : list.size;
            if (idx < 0 || idx > size) {
                throw new Error('position is out of range');
            }
        }
        else {
            switch (position) {
                case 'first':
                    idx = 0;
                    break;
                case 'last':
                    idx = Array.isArray(list) ? list.length : list.size;
                    break;
                case 'before':
                    idx = list.indexOf(reference);
                    if (idx === -1) {
                        throw new Error('reference not contained in this list');
                    }
                    break;
                case 'after':
                    idx = list.indexOf(reference) + 1;
                    if (idx === 0) {
                        throw new Error('reference not contained in this list');
                    }
                    break;
                default:
                    throw Error("Invalid position \"" + position + "\"");
            }
        }
        return idx;
    }
    function insertInList(list, item, position, reference) {
        return list.insert(getIndex(list, item, position, reference), item);
    }
    exports.insertInList = insertInList;
    function insertInArray(array, item, position, reference) {
        array.splice(getIndex(array, item, position, reference), 0, item);
        return array;
    }
    exports.insertInArray = insertInArray;
    function valueReplacer(key, value) {
        if (value instanceof RegExp) {
            return ("__RegExp(" + value.toString() + ")");
        }
        return value;
    }
    function valueReviver(key, value) {
        if (value.toString().indexOf('__RegExp(') === 0) {
            var _a = value.match(/__RegExp\(([^\)]*)\)/), regExpStr = _a[1];
            var _b = regExpStr.match(/^\/(.*?)\/([gimy]*)$/), regExp = _b[1], flags = _b[2];
            return new RegExp(regExp, flags);
        }
        return value;
    }
    /**
     * Internal function to convert a state value to a string
     * @param value The value to be converted
     */
    function valueToString(value) {
        return value
            ? Array.isArray(value) || typeof value === 'object'
                ? JSON.stringify(value, valueReplacer) : String(value)
            : value === 0
                ? '0' : value === false
                ? 'false' : '';
    }
    exports.valueToString = valueToString;
    /**
     * Internal function to convert a string to the likely more complex value stored in
     * state
     * @param str The string to convert to a state value
     */
    function stringToValue(str) {
        try {
            var value = JSON.parse(str, valueReviver);
            return value;
        }
        catch (e) {
            if (/^(\-|\+)?([0-9]+(\.[0-9]+)?|Infinity)$/.test(str)) {
                return Number(str);
            }
            if (str) {
                return str;
            }
            return undefined;
        }
    }
    exports.stringToValue = stringToValue;
});
