// dojo-widgets/dist/umd/main.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", './createButton', './createContainer', './createDijit', './createLayoutContainer', './createList', './createPanel', './createResizePanel', './createTabbedPanel', './createTextInput', './createWidget', './projector'], factory);
    }
})(function (require, exports) {
    "use strict";
    var createButton_1 = require('./createButton');
    exports.createButton = createButton_1.default;
    var createContainer_1 = require('./createContainer');
    exports.createContainer = createContainer_1.default;
    var createDijit_1 = require('./createDijit');
    exports.createDijit = createDijit_1.default;
    var createLayoutContainer_1 = require('./createLayoutContainer');
    exports.createLayoutContainer = createLayoutContainer_1.default;
    var createList_1 = require('./createList');
    exports.createList = createList_1.default;
    var createPanel_1 = require('./createPanel');
    exports.createPanel = createPanel_1.default;
    var createResizePanel_1 = require('./createResizePanel');
    exports.createResizePanel = createResizePanel_1.default;
    var createTabbedPanel_1 = require('./createTabbedPanel');
    exports.createTabbedPanel = createTabbedPanel_1.default;
    var createTextInput_1 = require('./createTextInput');
    exports.createTextInput = createTextInput_1.default;
    var createWidget_1 = require('./createWidget');
    exports.createWidget = createWidget_1.default;
    var projector_1 = require('./projector');
    exports.projector = projector_1.default;
});
